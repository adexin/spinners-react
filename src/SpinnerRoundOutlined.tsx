import React from 'react';
import { defaultProps, SpinnerProps } from './helpers';
import { SpinnersProps, withSharedProps } from './withSharedProps';

import './SpinnerRoundOutlined.css';

const animations = [
  {
    r: 2,
  },
  {
    name: 'spinners-react-round-outlined',
    r: 14,
  },
  {
    name: 'spinners-react-round-outlined',
    r: 28,
  },
];

export type SpinnerRoundOutlinedProps = SpinnersProps & SpinnerProps;

const Component = ({
  speed = defaultProps.speed,
  still = defaultProps.still,
  thickness = defaultProps.thickness,
  ...svgProps
}: SpinnerProps) => {
  const strokeWidth = 3 * (thickness / 100);

  return (
    <svg fill="none" {...svgProps} viewBox="0 0 66 66">
      {animations.map((animation, i) => (
        <circle
          key={`spinner-round-outlined-r${animation.r}`}
          cx="33"
          cy="33"
          fill="none"
          r={animation.r}
          stroke="currentColor"
          strokeWidth={i ? strokeWidth : 4}
          style={animation.name && !still
            ? { animation: `${animation.name} ${140 / speed}s ease-in-out infinite` }
            : {}}
        />
      ))}
    </svg>
  );
};

export const SpinnerRoundOutlined = withSharedProps(Component);
