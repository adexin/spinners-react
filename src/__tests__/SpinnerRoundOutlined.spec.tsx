import { render } from '@testing-library/react';
import React from 'react';

import { SpinnerRoundOutlined } from '../SpinnerRoundOutlined';

describe('SpinnerRoundOutlined', () => {
  const color = 'red';
  const animations = [
    {
      r: 2,
      strokeWidth: 4,
    },
    {
      name: 'spinners-react-round-outlined',
      r: 14,
    },
    {
      name: 'spinners-react-round-outlined',
      r: 28,
    },
  ];

  it('matches the snapshot', () => {
    const { container } = render(
      <SpinnerRoundOutlined color="#fff" size={100} speed={10} thickness={50} />,
    );

    expect(container).toMatchSnapshot();
  });

  it('matches the snapshot with default props', () => {
    const { container } = render(<SpinnerRoundOutlined />);

    expect(container).toMatchSnapshot();
  });

  it('renders null if disabled', () => {
    const { container } = render(<SpinnerRoundOutlined enabled={false} />);

    expect(container.firstChild).toBe(null);
  });

  it('renders still if specified', () => {
    const { container } = render(<SpinnerRoundOutlined still />);
    const circles = container.querySelectorAll('circle');

    circles.forEach((circle) => {
      expect(circle.style.animation).toBe('');
    });
  });

  it('passes props to styles', () => {
    const size = '20%';
    const thickness = 40;
    const speed = 50;
    const { container } = render(
      <SpinnerRoundOutlined
        color={color}
        size={size}
        speed={speed}
        thickness={thickness}
      />,
    );
    const circles = container.querySelectorAll('circle');

    expect(container.firstChild).toHaveStyle({ width: size });
    circles.forEach((circle, index) => {
      if (index) {
        expect(circle.style.animation.includes(`${140 / speed}`)).toBe(true);
        expect(circle.style.animation.includes(animations[index].name)).toBe(
          true,
        );
        expect(circle).toHaveAttribute(
          'stroke-width',
          `${3 * (thickness / 100)}`,
        );
      } else {
        expect(circle).toHaveAttribute(
          'stroke-width',
          `${animations[index].strokeWidth}`,
        );
        expect(circle.style.animation).toBe('');
      }
    });
  });

  it('passes svg props overriding styles', () => {
    const className = 'test-class';
    const { container } = render(
      <SpinnerRoundOutlined
        className={className}
        color="green"
        style={{ color }}
      />,
    );

    expect(container.firstChild).toHaveClass(className);
    expect(container.firstChild).toHaveStyle({ color });
  });
});
