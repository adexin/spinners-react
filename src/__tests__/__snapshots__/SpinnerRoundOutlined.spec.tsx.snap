// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpinnerRoundOutlined matches the snapshot 1`] = `
<div>
  <svg
    fill="none"
    style="color: rgb(255, 255, 255); overflow: visible; width: 100px;"
    viewBox="0 0 66 66"
  >
    <circle
      cx="33"
      cy="33"
      fill="none"
      r="2"
      stroke="currentColor"
      stroke-width="4"
    />
    <circle
      cx="33"
      cy="33"
      fill="none"
      r="14"
      stroke="currentColor"
      stroke-width="1.5"
      style="animation: spinners-react-round-outlined 14s ease-in-out infinite;"
    />
    <circle
      cx="33"
      cy="33"
      fill="none"
      r="28"
      stroke="currentColor"
      stroke-width="1.5"
      style="animation: spinners-react-round-outlined 14s ease-in-out infinite;"
    />
  </svg>
</div>
`;

exports[`SpinnerRoundOutlined matches the snapshot with default props 1`] = `
<div>
  <svg
    fill="none"
    style="color: rgb(56, 173, 72); overflow: visible; width: 50px;"
    viewBox="0 0 66 66"
  >
    <circle
      cx="33"
      cy="33"
      fill="none"
      r="2"
      stroke="currentColor"
      stroke-width="4"
    />
    <circle
      cx="33"
      cy="33"
      fill="none"
      r="14"
      stroke="currentColor"
      stroke-width="3"
      style="animation: spinners-react-round-outlined 1.4s ease-in-out infinite;"
    />
    <circle
      cx="33"
      cy="33"
      fill="none"
      r="28"
      stroke="currentColor"
      stroke-width="3"
      style="animation: spinners-react-round-outlined 1.4s ease-in-out infinite;"
    />
  </svg>
</div>
`;
